cluster-up/cluster/k3d/common.sh

```diff
@@ -116,38 +120,67 @@ function _label_agents() {
     done
 }
 
+function _cluster_nodes_args() {
+    args="--servers=$KUBEVIRT_NUM_SERVERS --agents=$KUBEVIRT_NUM_AGENTS "
+    for i in $(seq 0 $((KUBEVIRT_NUM_SERVERS - 1))); do
+        id=${KUBEVIRTCI_CONFIG_PATH}/$KUBEVIRT_PROVIDER/machine-server-id-$i
+        printf "%.0s$((i + 1))" {1..32} > ${id}
+        args=${args}"-v ${id}:/etc/machine-id@server:${i} "
+    done
+
+    for j in $(seq 0 $((KUBEVIRT_NUM_AGENTS - 1))); do
+        id=${KUBEVIRTCI_CONFIG_PATH}/$KUBEVIRT_PROVIDER/machine-agent-id-$j
+        printf "%.0s$(($KUBEVIRT_NUM_SERVERS + $j + 1))" {1..32} > ${id}
+        args=${args}"-v ${id}:/etc/machine-id@agent:${j} "
+    done
+    echo "${args}"
+}
+
+function _cni_args() {
+    CALICO=$(pwd)/cluster-up/cluster/k3d/manifests/calico.yaml
+    if [ ${K3D_CNI} == calico ]; then
+       args=" --k3s-arg --flannel-backend=none@server:* \
+-v $CALICO:/var/lib/rancher/k3s/server/manifests/calico.yaml@server:* "
+    fi
+    echo "${args}"
+}
+
+function _device_args() {
+    if [ ${VFIO_ENABLED} == true ]; then
+        args=" -v /dev/vfio:/dev/vfio@agent:* "
+    fi
+    echo "${args}"
+}
+
+function _generate_k3d_args() {
+    args=$(_cluster_nodes_args)$(_cni_args)$(_device_args)
+    if [ ${DISABLE_DEFAULT_SERVICES} == true ] ; then
+        args=${args}" --k3s-arg --disable=traefik,servicelb,metrics-server@server:* "
+    else
+        args=${args}" --k3s-arg --disable=traefik@server:0 "
+    fi
+    args=${args}" --registry-use $REGISTRY_NAME \
+--api-port $KUBERNETES_SERVICE_HOST:$KUBERNETES_SERVICE_PORT \
+--no-lb \
+--k3s-arg "--kubelet-arg=cpu-manager-policy=static@agent:*" \
+--k3s-arg "--kubelet-arg=kube-reserved=cpu=500m@agent:*" \
+--k3s-arg "--kubelet-arg=system-reserved=cpu=500m@agent:*" \
+-v /lib/modules:/lib/modules@agent:* \
+"
+    echo "${args}"
+}
+
+
 function _create_cluser() {
```

Comment:
Possible idea until we have manifests,
add a function to `common.sh` `named _create_arm_cluster`  which has all the changes needed
and then according the ARCH you can know which function to trigger
wdyt ?

Reply:
If this script is only meant to serve CI/CD purposes, then it should suffice. However, if it's intended for use by normal users, it may lead to confusion since there is no CPU arch specific setting, but rather a naming convention with _create_arm_cluster.

In my opinion, if we want this script to have wider usage, it should be constructed like building blocks. We can have the basic parts in common.sh and put specific settings in another folder. In other words, it should be able to create a basic k3d cluster, and any other specific purposes can be built on top of it.

The core issue is figuring out which parts should be included in the basic k3d cluster and which should be put into the specific folders.

Reply:
How about make this simple, just have a environmental variable `K3D_SINGLE_NODE` which can be true or false. If it is set to true, apply the adapted setting for Arm64 E2E testing, otherwise use sriov setting?

Here is an issue, if the E2E testing for Arm64 needs multi-node cluster, we need to change the script again.

Reply:
We use this lane for `SR-IOV` mostly and should
allow local development which just spin a cluster without `SR-IOV`
so the non `SR-IOV` should allow more than one node in the future, and not just `ARM` as you raised.
We can of course consider doing it simple and in steps.

I prefer that we continue next week please, after our vacation,
won't manage to review it before.
Thanks

Reply:
> I prefer that we continue next week please, after our vacation, won't manage to review it before. Thanks

Ok, no problem. Thanks @oshoval